### PR TITLE
Kindle: toggle cover events

### PR DESCRIPTION
**Pull request by yparitcher**
~~used on kindle but can be extended to other devices~~

Allow disabling the hall efect sensor via the sysfs knob, so the kindle system wont sleep & wake the device

for those of use that stay in koreader, are caseless and get spurious wakeups

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10363)

### CODE REVIEW

**Comment by NiLuJe** [comment]
I don't quite recall how this works on Kindle, but, FWIW, on Kobo, we essentially get an EV_KEY with a specific keycode, so we can easily ignore the event (and go back to sleep if necessary).

It's handled over here for the UI side:

https://github.com/koreader/koreader/blob/6c2f2306393e82dced8867da9eceb9e0be3a536d/frontend/ui/elements/common_settings_menu_table.lua#L213-L237

And sorted somewhere in Device:

https://github.com/koreader/koreader/blob/6c2f2306393e82dced8867da9eceb9e0be3a536d/frontend/ui/elements/filemanager_menu_order.lua#L59

----

TL;DR: Feel free to move the menu entry somewhere around there, too?

File: frontend/ui/elements/common_settings_menu_table.lua [hunk 1]
```diff
@@ -236,6 +237,15 @@ if Device:isKobo() then
     }
 end
 
+if Device:isKindle() and PowerD:hasHallSensor() then
+    common_settings.hall_sensor = {
+        text = _("Hall Sensor"),
```
Comment (poire-z) [inline frontend/ui/elements/common_settings_menu_table.lua:242]: You probably need a help_text:
`Detect when you are in a building or large room used for events involving a lot of people, and do nothing.`

(unlesss I didn't get what this does right :))

**Comment by poire-z** [comment]
Oh, ok, it's just sleep cover handling ! :)
I guess a menu item "Hall sensor", even with your help_text, is too technical: just name it like the existing menu for handling the sleep cover on Kobo ?
Or better, I guess it should be abstracted low level so the same menu and same bit of code handle like like they do on a Kobo? (unless it's really too different, and we don't have to handle anything at all if it's the Kindle OS that do it all without us even knowing?)

**Comment by Frenzie** [comment]
We already have code at least in Kobo to deal with spurious wakeups, basically a reasonably quick go back to sleep if no input for a minute (or whatever). As for the menu item I think it should probably just reuse the Kobo "sleepcover event" message; the technical side you'd have to check what the best way is to abstract it away.

**Comment by yparitcher** [comment]
As usual on kindle we cant get the events, we can only change the sysfs knob to enable / disable on a driver level

**Comment by yparitcher** [comment]
Kindle also does not advertise as a sleep cover so kindle users might not be familiar with the term. Maybe cover events?

Also i need help with `device_spec` i am not sure how to convert its expecting us to use os.execute when we don't use it anymore

**Comment by Frenzie** [comment]
I think that maps fairly cleanly to "ignore all sleepcover events"? :-)

**Comment by Frenzie** [comment]
> Kindle also does not advertise as a sleep cover so kindle users might not be familiar with the term. Maybe cover events?

What term does Kindle use?

**Comment by yparitcher** [comment]
> > Kindle also does not advertise as a sleep cover so kindle users might not be familiar with the term. Maybe cover events?
> 
> What term does Kindle use?

Kindle {model} cover

Deep in product details it says wakes and sleeps kindle

File: frontend/util.lua [hunk 2]
```diff
@@ -829,6 +833,26 @@ function util.removeFile(file)
     end
 end
 
+function util.writeToSysfs(val, file)
+    -- NOTE: We do things by hand via ffi, because io.write uses fwrite,
+    --       which isn't a great fit for procfs/sysfs (e.g., we lose failure cases like EBUSY,
+    --       as it only reports failures to write to the *stream*, not to the disk/file!).
+    local fd = C.open(file, bit.bor(C.O_WRONLY, C.O_CLOEXEC)) -- procfs/sysfs, we shouldn't need O_TRUNC
```
[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/util.lua:840]: We also dont have the hall_file path on any legacy kindles so should not be an issue in practice

[the remark this answers is not in the document]
Comment (NiLuJe) [reply frontend/util.lua:840]: Oh, duh' ;o).

[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/util.lua:840]: Can you help with the tests? `device_spec` is failing, and i am not sure how to convert its expecting us to use os.execute when we don't use it anymore. line 216

[the remark this answers is not in the document]
Comment (NiLuJe) [reply frontend/util.lua:840]: Is the "fuck it" approach enough to appease it?

```diff
diff --git a/spec/unit/device_spec.lua b/spec/unit/device_spec.lua
index 7302ee4f5..7e5abb025 100644
--- a/spec/unit/device_spec.lua
+++ b/spec/unit/device_spec.lua
@@ -40,7 +40,6 @@ describe("device module", function()
     after_each(function()
         mock_input.open:revert()
         os.getenv:revert()
-        os.execute:revert()
 
         os.getenv = osgetenv
         io.open = iopen
@@ -213,8 +212,6 @@ describe("device module", function()
             assert.is.same(kindle_dev.powerd.fl_intensity, 5)
 
             kindle_dev.powerd:toggleFrontlight()
-            assert.stub(os.execute).was_called_with(
-                "printf '%s' 0 > /sys/class/backlight/max77696-bl/brightness")
             -- Here be shenanigans: we don't override powerd's fl_intensity when we turn the light off,
             -- so that we can properly turn it back on at the previous intensity ;)
             assert.is.same(kindle_dev.powerd.fl_intensity, 5)
```

(It's not like we could stub the new FFI call anyway).

[the remark this answers is not in the document]
Comment (yparitcher) [reply frontend/util.lua:840]: That was my plan. i just wanted to see if you had a way to salvage the wreckage :)

[the remark this answers is not in the document]
Comment (NiLuJe) [reply frontend/util.lua:840]: Besides going full tilt overengineering and stubbing `util.writeToSys` altogether (which would barely make sense), nope; so you've got my blessing ^^.

**Comment by NiLuJe** [comment]
Random fun fact: I seem to recall that (on the PW5, at least) there's actually two of them, one calibrated for the back, and one for the front. This is (I imagine) to help deal with tolerance issues to prevent false-positives from the back (when there's an actual cover, at least).

I don't recall how/whel/why I even checked that, because I've never used a cover with a Kindle, but I remember thinking that it looked neat (especially given the amount of crap false positives like that have caused on kobo ;p).

File: frontend/ui/elements/common_settings_menu_table.lua [hunk 3]
```diff
@@ -236,6 +237,17 @@ if Device:isKobo() then
     }
 end
 
+if Device:isKindle() and PowerD:hasHallSensor() then
+    common_settings.cover_events = {
+        text = _("Cover events"),
+        help_text = _([[Enable or disable the hall effect sensor.
```
Comment (hius07) [inline frontend/ui/elements/common_settings_menu_table.lua:243]: "The Hall-effect principle is named for physicist Edwin Hall. In 1879 he discovered that when a conductor or semiconductor with current flowing in one direction was introduced perpendicular to a magnetic field a voltage could be measured at right angles to the current path."
May be capitilize "hall"?

Comment (Frenzie) [reply frontend/ui/elements/common_settings_menu_table.lua:243]: Incidentally, I've been considering buying a [Gulikit KingKong 2](https://www.gulikit.com/productinfo/737791.html), with Hall effect sensors.

But over here as I said I'm not sure how much sense it makes to talk about the underlying tech instead of what it does (i.e., detect covers).

Comment (poire-z) [reply frontend/ui/elements/common_settings_menu_table.lua:243]: Also, I understand that the Kindle has it enabled by default ? And that you added that option to allow disabling it because it sucks ?
Then I would just name it `[ ] Disable cover events` (or `[ ] Disable Kindle cover events` if it's really clear to all it's Kindle system stuff outside of us).

Comment (hius07) [reply frontend/ui/elements/common_settings_menu_table.lua:243]: I meant the capitalization only.

File: frontend/ui/elements/common_settings_menu_table.lua [hunk 4]
```diff
@@ -236,6 +237,17 @@ if Device:isKobo() then
...
+This is used to detect if the cover is closed to auto sleep and wake the device. If there is no cover present the sensor may cause spurious wakeups when located next to a magnetic source.]]),
+        keep_menu_open = true,
+        checked_func = function() return PowerD:getHallSensor() end,
```
Comment (poire-z) [inline frontend/ui/elements/common_settings_menu_table.lua:246]: `PowerD:isHallSensorEnabled()` ?

**Comment by yparitcher** [comment]
> Random fun fact: I seem to recall that (on the PW5, at least) there's actually (at least?) two of them, one calibrated for the back, and one for the front. This is (I imagine) to help deal with tolerance issues to prevent false-positives from the back (when there's an actual cover involved, at least).
> 
> I don't recall how/where/why I even checked that, because I've never used a cover with a Kindle, but I remember thinking that it the implementation looked neat (especially given the amount of crap false positives like that have caused on kobo ;p).

Makes sense, as it is more difficult to trigger this on my pw5 then on older models

**Comment by yparitcher** [comment]
Ready to merge

**Review by Frenzie — APPROVED** [review]
lgtm, left a small phrasing nit
